docs/05-engineering/git-gitlab-github-bitbucket.md

```diff
@@ -1,29 +1,144 @@
-# Git, GitLab, GitHub, and Bitbucket
+# Git
 
 ## Git Overview
 
-As you probably already know, Git is a source code versioning system that lets you locally track changes and push or pull changes from remote resources. Many projects use Git. Here are some references to review if you are unfamiliar with it.
+As you probably already know, Git is a source code versioning system that lets you locally track changes and push or pull changes from remote resources. All projects use Git. Here are some references to review if you are unfamiliar with it.
 
 - <https://git-scm.com/>
 - <https://help.github.com/articles/git-and-github-learning-resources/>
 - Getting Git Right <https://www.atlassian.com/git/>
 - Version Control with Git book <http://www.amazon.com/Version-Control-Git-collaborative-development/dp/1449316387>
 - Learn "how git works" at a lower level, and some useful stuff like rebasing (workflow and interactive), reflog, etc. <https://www.youtube.com/watch?v=MYP56QJpDr4>
-- if you want a non-video companion to the first half, <http://git-scm.com/book/en/v2/Git-Internals-Git-Objects> is a great place to start.
+- If you want a non-video companion to the first half, <http://git-scm.com/book/en/v2/Git-Internals-Git-Objects> is a great place to start.
 
-## What's the difference between GitLab, GitHub, and Bitbucket?
+## Git best practices
+
+_Note: your project may include additional or alternate code review practices - please review your project developer/contributor documentation!_
+
+- Atomic commits:
+  - Generally commits should be as small as possible.
+  - The code should be in a functional state after each commit (even if incomplete) - avoid committing code which will won't run at all.
+- Write good commit messages. A good commit message should:
+  - Reference the associated ticket ID using a prefix, e.g. `ABC-123:`.
```

Comment:
I haven't done this for a while. Generally you have one PR per ticket, so the ticket number in the branch makes more sense and saves on some redundancy. Many times we also skip the ticket name from the branch, and add it to the PR comment, or simply add a link to a PR in whatever ticketing system we are using.

Reply:
Whilst I think ticket prefixes is certainly an optional thing on a project level (I don't bother on random personal projects for example!) - I think it's still a good practice and is the norm on most of our projects. It just makes it that much faster to go directly to the ticket from git logs or blame output, rather than having to figure out what PR/MR originated it and then follow links from there (hopefully, if someone remembered to add one) back to the ticket - not bad for 1 change, but if you have a dozen changes touching a function and want to find the tickets for all of them it can take a while.
If applied to a pre-commit and/or CI check it also strongly discourages "random" work that isn't associated with a ticket - go make a quick ticket first even if it's just a placeholder while the idea is fresh in your head.
I'll update this with a little more detail on the purpose of this at least.